### PR TITLE
More flexible selection of vertical and horizontal channels

### PR DESCRIPTION
**Pull request by flixha**
### What does this PR do?
Allows user to choose freely which channels should be associated with P-waves ("vertical channels") and which should be associated with S-waves ("horizontal channels"), and whether picks on one of the vertical channels should be treated to apply to all channels in that group (so like `all_horiz`).

### Why was it initiated?  Any relevant Issues?
While EQcorrscan currently has the option to choose parameters `all_horiz` and to define the horizontal channels with `horizontal_chans`, there is no equivalent option for the channels primarily associated with P-picks. This can be relevant e.g., when for an OBS deployment, a user wants to have Z- and D-channels (vertical, pressure) associated with the P-wave, and  N/E (or 1/2) associated with S-arrivals. For simplicity let's call Z and D the "vertical channels", even though the hydrophone isn't really that.

The options were already in place in some functions, but it wasn't working fully for all relevant functions, i.e., `template_gen`, `lag_calc`, and some helper functions.

Remaining points:
~- be consistent with default values for `horizontal_chans`, which appears both as `['E', 'N', '1', '2', '3']` and `['E', 'N', '1', '2']`~
~- add full test for functionality~

### PR Checklist
- [X] `develop` base branch selected?
- [X] This PR is not directly related to an existing issue (which has no PR yet).
- [X] All tests still pass.
- [X] Any new features or fixed regressions are be covered via new tests.
- [X] Any new or changed features have are fully documented.
- [X] Significant changes have been added to `CHANGES.md`.
~- [ ] First time contributors have added your name to `CONTRIBUTORS.md`.~

### CODE REVIEW

**Comment by flixha** [comment]
I now added a test case for ocean bottom seismometer data, with 4 channels (HDH and HHZ for P-waves), and with T-arrivals (that should also be handled on the vertical channels).

Tests on python 3.8 and 3.9 are all failing now, I suspect it has to do with a new Scipy version 1.9.0 - looking into that.

**Comment by flixha** [comment]
In Scipy 1.9.0; there is no more `'hanning` window accessible, instead it's only accessible as `'hann`.

**Comment by flixha** [comment]
In Scipy>1.9.0, there is no more `"hanning"` window; rather it's only accessible now as `"hann"`; see: https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.windows.hann.html

**Comment by flixha** [comment]
Tests still failing due to the same hanning-issue in obspy (see https://github.com/obspy/obspy/pull/3117), so I'm setting a limit on the Scipy-version (<1.9.0) for now until obspy is updated.

File: eqcorrscan/core/lag_calc.py [hunk 1]
```diff
@@ -396,7 +399,9 @@ def xcorr_pick_family(family, stream, shift_len=0.2, min_cc=0.4,
     return picked_dict
 
 
-def _prepare_data(family, detect_data, shift_len):
+def _prepare_data(family, detect_data, shift_len, all_vert=False,
+                  all_horiz=False, vertical_chans=['Z'],
+                  horizontal_chans=['E', 'N', '1', '2']):
```
Comment (calum-chamberlain) [inline eqcorrscan/core/lag_calc.py:404]: This probably won't come up as an issue because this is a private function, but it might be better if `horizontal_chans` is defined as the same default as in `xcorr_pick_family`.

Comment (calum-chamberlain) [reply eqcorrscan/core/lag_calc.py:404]: Actually, this is defined in a few places - if you don't have a reason for them not being consistent (it may be legacy from my inconsistency?)  then we should make them the same.

File: eqcorrscan/core/lag_calc.py [hunk 2]
```diff
@@ -221,8 +221,9 @@ def _concatenate_and_correlate(streams, template, cores):
 
 def xcorr_pick_family(family, stream, shift_len=0.2, min_cc=0.4,
                       min_cc_from_mean_cc_factor=None,
-                      horizontal_chans=['E', 'N', '1', '2'],
-                      vertical_chans=['Z'], cores=1, interpolate=False,
+                      all_vert=False, all_horiz=False, vertical_chans=['Z'],
+                      horizontal_chans=['E', 'N', '1', '2', '3'],
```
Comment (calum-chamberlain) [inline eqcorrscan/core/lag_calc.py:225]:
```suggestion
                      horizontal_chans=['E', 'N', '1', '2'],
```

File: eqcorrscan/core/match_filter/detection.py [hunk 3]
```diff
@@ -348,7 +348,9 @@ def _calculate_event(self, template=None, template_st=None,
         self.event = ev
         return self
 
-    def extract_stream(self, stream, length, prepick):
+    def extract_stream(self, stream, length, prepick, all_vert=False,
+                       all_horiz=False, vertical_chans=['Z'],
+                       horizontal_chans=['E', 'N', '1', '2', '3']):
```
Comment (calum-chamberlain) [inline eqcorrscan/core/match_filter/detection.py:353]:
```suggestion
                       horizontal_chans=['E', 'N', '1', '2']):
```

File: eqcorrscan/core/match_filter/family.py [hunk 4]
```diff
@@ -507,8 +507,8 @@ def write(self, filename, format='tar', overwrite=False):
         return
 
     def lag_calc(self, stream, pre_processed, shift_len=0.2, min_cc=0.4,
-                 min_cc_from_mean_cc_factor=None,
-                 horizontal_chans=['E', 'N', '1', '2'], vertical_chans=['Z'],
+                 min_cc_from_mean_cc_factor=None, vertical_chans=['Z'],
+                 horizontal_chans=['E', 'N', '1', '2', '3'],
```
Comment (calum-chamberlain) [inline eqcorrscan/core/match_filter/family.py:511]:
```suggestion
                 horizontal_chans=['E', 'N', '1', '2'],
```

File: eqcorrscan/core/match_filter/family.py [hunk 5]
```diff
@@ -779,7 +779,9 @@ def _process_streams(self, stream, pre_processed, process_cores=1,
             processed_stream = stream.merge()
         return processed_stream.split()
 
-    def extract_streams(self, stream, length, prepick):
+    def extract_streams(self, stream, length, prepick, all_vert=False,
+                        all_horiz=False, vertical_chans=['Z'],
+                        horizontal_chans=['E', 'N', '1', '2', '3']):
```
Comment (calum-chamberlain) [inline eqcorrscan/core/match_filter/family.py:784]:
```suggestion
                        horizontal_chans=['E', 'N', '1', '2']):
```

File: eqcorrscan/core/match_filter/party.py [hunk 6]
```diff
@@ -833,8 +833,8 @@ def read(self, filename=None, read_detection_catalog=True,
         return self
 
     def lag_calc(self, stream, pre_processed, shift_len=0.2, min_cc=0.4,
-                 min_cc_from_mean_cc_factor=None,
-                 horizontal_chans=['E', 'N', '1', '2'], vertical_chans=['Z'],
+                 min_cc_from_mean_cc_factor=None, vertical_chans=['Z'],
+                 horizontal_chans=['E', 'N', '1', '2', '3'],
```
Comment (calum-chamberlain) [inline eqcorrscan/core/match_filter/party.py:837]:
```suggestion
                 horizontal_chans=['E', 'N', '1', '2'],
```

File: eqcorrscan/core/template_gen.py [hunk 7]
```diff
@@ -52,10 +52,11 @@ def __str__(self):
...
                  length, prepick, swin="all", process_len=86400,
-                 all_horiz=False, delayed=True, plot=False, plotdir=None,
-                 return_event=False, min_snr=None, parallel=False,
-                 num_cores=False, save_progress=False, skip_short_chans=False,
-                 **kwargs):
+                 all_vert=False, all_horiz=False, delayed=True, plot=False,
+                 plotdir=None, return_event=False, min_snr=None,
+                 parallel=False, num_cores=False, save_progress=False,
+                 skip_short_chans=False, vertical_chans=['Z'],
+                 horizontal_chans=['E', 'N', '1', '2', '3'], **kwargs):
```
Comment (calum-chamberlain) [inline eqcorrscan/core/template_gen.py:59]:
```suggestion
                 horizontal_chans=['E', 'N', '1', '2'], **kwargs):
```

File: eqcorrscan/core/template_gen.py [hunk 8]
```diff
@@ -582,9 +596,10 @@ def _rms(array):
     return np.sqrt(np.mean(np.square(array)))
 
 
-def _template_gen(picks, st, length, swin='all', prepick=0.05,
+def _template_gen(picks, st, length, swin='all', prepick=0.05, all_vert=False,
                   all_horiz=False, delayed=True, plot=False, min_snr=None,
-                  plotdir=None):
+                  plotdir=None, vertical_chans=['Z'],
+                  horizontal_chans=['E', 'N', '1', '2', '3']):
```
Comment (calum-chamberlain) [inline eqcorrscan/core/template_gen.py:602]:
```suggestion
                  horizontal_chans=['E', 'N', '1', '2']):
```

**Review by calum-chamberlain — CHANGES_REQUESTED** [review]
Can you either go with 1, 2, 3 or 1, 2 as defaults, then we should merge this. I have tried to catch all the cases in the changed lines, but there may be more.

**Review by calum-chamberlain — APPROVED** [review]

**Review by calum-chamberlain — APPROVED** [review]

**Comment by calum-chamberlain** [comment]
Looks good to me, and I don't see the coverage complaints - maybe they were raised before the full test suite was finished? Looks good to me so I'm going to merge. Thanks for adding that edge case test.

As an aside, that log handler looks like it is probably handy. If you found that helpful in debugging tests you might consider adding a wrapper that can be added to all tests to set up logging for them.
